webui: introduce OpenAI-compatible model selector in JSON payload

ServeurpersoCom opened the pull request:
# Introduce OpenAI-compatible model selector in JSON payload

This PR adds a minimal model selector to the WebUI sidebar, allowing users to pick an available model exposed through the /v1/models OpenAI-compatible endpoint

The selector automatically fetches and lists models from the server, persists the selected model in local storage, and sends it in the JSON body of subsequent /v1/chat/completions requests. The selection logic mirrors OpenAI’s client behavior while remaining fully offline-compatible with local llama.cpp instances

This enables direct interoperability with OpenAI-compatible clients and simplifies multi-model setups in the WebUI

## Restore OpenAI-Compatible model source of truth and unify metadata capture :

This change re-establishes a single, reliable source of truth for the active model:
fully aligned with the OpenAI-Compat API behavior

It introduces a unified metadata flow that captures the model field from both
streaming and non-streaming responses, wiring a new onModel callback through ChatService
The model name is now resolved directly from the API payload rather than relying on
server /props or UI assumptions

ChatStore records and persists the resolved model for each assistant message during
streaming, ensuring consistency across the UI and database
Type definitions for API and settings were also extended to include model metadata
and the onModel callback, completing the alignment with OpenAI-Compat semantics

## Remaining '/props' usage audit in the WebUI :

A repository-wide search inside 'tools/server/webui' shows the remaining '/props' references are intentional because the WebUI still needs to bootstrap and validate server capabilities outside of chat responses:

* 'src/routes/+layout.svelte' and 'src/lib/stores/server.svelte.ts' fetch '/props' on application startup to populate the global server store with template, model alias, and capability metadata that never appears in chat completions.
* 'src/lib/components/app/server/ServerErrorSplash.svelte' and 'src/lib/components/app/chat/ChatScreen/ChatScreenWarning.svelte' surface fallback UI when '/props' is unreachable, ensuring the user understands cached data might be stale.
* 'src/lib/utils/api-key-validation.ts' validates API keys against '/props' so that the UI can warn about incompatible keys before issuing chat requests.
* 'src/lib/services/chat.ts' performs a last-resort fetch to '/props' when the streaming handshake fails, preserving compatibility with legacy servers that only expose model names via that endpoint.

ServeurpersoCom committed:
webui: introduce OpenAI-compatible model selector in JSON payload

ServeurpersoCom committed:
webui: restore OpenAI-Compatible model source of truth and unify metadata capture

This change re-establishes a single, reliable source of truth for the active model:
fully aligned with the OpenAI-Compat API behavior

It introduces a unified metadata flow that captures the model field from both
streaming and non-streaming responses, wiring a new onModel callback through ChatService
The model name is now resolved directly from the API payload rather than relying on
server /props or UI assumptions

ChatStore records and persists the resolved model for each assistant message during
streaming, ensuring consistency across the UI and database
Type definitions for API and settings were also extended to include model metadata
and the onModel callback, completing the alignment with OpenAI-Compat semantics

ServeurpersoCom commented:
@ngxson :) What do you think about this approach ?

- aiming to stay compatible with the current standalone llama-server,
- llama-swap
- and future multi-model evolutions of llama-server?

It introduces a unified, KISS, OpenAI-compatible model selection path while keeping everything backward-compatible with existing setups

allozaur commented on tools/server/webui/src/routes/+layout.svelte:
```diff
@@ -139,7 +139,7 @@
 <ModeWatcher />
 
-<Toaster richColors />
+<Toaster position="top-center" richColors />
```
Not sure about this... I feel that the toaster notifications are much better in their own space at their default bottom right corner as there's no other UI that they might be obfuscating when it's not necessary.
```suggestion
<Toaster richColors />
```

allozaur commented on tools/server/webui/src/lib/stores/models.svelte.ts:
```diff
@@ -0,0 +1,223 @@
+	model: string;
+};
+
+const STORAGE_KEY = 'llama.cpp:selectedModel';
```
Let's move this to `$lib/constants`

allozaur commented on tools/server/webui/src/lib/stores/models.svelte.ts:
```diff
@@ -0,0 +1,223 @@
+			const persisted = this.readPersistedSelection();
+			let nextSelectionId = this._selectedModelId ?? persisted?.id ?? null;
+			let nextSelectionName = this._selectedModelName ?? persisted?.model ?? null;
+			if (nextSelectionId) {
```
```suggestion
			let nextSelectionName = this._selectedModelName ?? persisted?.model ?? null;

			if (nextSelectionId) {
```

allozaur commented on tools/server/webui/src/lib/stores/models.svelte.ts:
```diff
@@ -0,0 +1,223 @@
+			let nextSelectionName = this._selectedModelName ?? persisted?.model ?? null;
+			if (nextSelectionId) {
+				const match = models.find((model) => model.id === nextSelectionId);
+				if (match) {
```
```suggestion
				const match = models.find((model) => model.id === nextSelectionId);

				if (match) {
```

allozaur commented on tools/server/webui/src/lib/stores/models.svelte.ts:
```diff
@@ -0,0 +1,223 @@
+				const id = parsed.id;
+				const model =
+					typeof parsed.model === 'string' && parsed.model.length > 0 ? parsed.model : id;
+				return { id, model };
```
```suggestion
					typeof parsed.model === 'string' && parsed.model.length > 0 ? parsed.model : id;

				return { id, model };
```

allozaur commented on tools/server/webui/src/lib/stores/models.svelte.ts:
```diff
@@ -0,0 +1,223 @@
+			const models: ModelOption[] = response.data.map((item, index) => {
+				const details = response.models?.[index];
+				const rawCapabilities = Array.isArray(details?.capabilities)
+					? [...(details?.capabilities ?? [])]
```
This is already checked with `Array.isArray(details?.capabilities)`, so there's no need for additional null coalescing here.

```suggestion
					? details?.capabilities
```

ServeurpersoCom committed:
webui: address review feedback from allozaur

ServeurpersoCom commented:
Extracted determineInitialSelection helper
Centralized localStorage key constant
Minor Toaster cleanup per review
All checks passing locally

ServeurpersoCom committed:
webui: move model selector into ChatForm (idea by @allozaur)

ServeurpersoCom commented:
> I think that placing the model selector in the Sidebar makes its UI a bit too heavy and bloated... Much better place for changing the model woud be in Chat Form, like here:

Computer :
<img width="1705" height="813" alt="PC1" src="https://github.com/user-attachments/assets/9a387d6f-ec47-41dd-bcb5-371c0542a732" />
<img width="1711" height="785" alt="PC2" src="https://github.com/user-attachments/assets/9a21f531-4ced-46bb-96f7-dacd6458ef19" />
<img width="1705" height="819" alt="PC3" src="https://github.com/user-attachments/assets/b8abb8ef-5883-4b22-b251-ebe034cb6565" />

Smartphone :
<img width="184" height="400" alt="Mobile1" src="https://github.com/user-attachments/assets/29193c8c-dd4b-4951-b915-4e2f7dc3503c" /> <img width="184" height="400" alt="Mobile2" src="https://github.com/user-attachments/assets/5005c2bd-373e-4b4c-9b4e-2e4df1c473ac" /> <img width="184" height="400" alt="Mobile3" src="https://github.com/user-attachments/assets/275ec48d-b25f-45f7-9efc-6d6fde6f22be" />

That’s actually a great idea : moving the selector into the Chat Form feels way more natural now that I’ve tested it 😄
The layout is cleaner, lighter, and it fits perfectly in the conversation flow. Definitely the right spot!

https://github.com/ggml-org/llama.cpp/pull/16562/commits/1324a4df92b0342a91fdc96660d1ae1704aa40ca

allozaur commented:
> > I think that placing the model selector in the Sidebar makes its UI a bit too heavy and bloated... Much better place for changing the model woud be in Chat Form, like here:
> 
> 
> 
> Computer :
> 
> <img width="1705" height="813" alt="PC1" src="https://github.com/user-attachments/assets/9a387d6f-ec47-41dd-bcb5-371c0542a732" />
> 
> <img width="1711" height="785" alt="PC2" src="https://github.com/user-attachments/assets/9a21f531-4ced-46bb-96f7-dacd6458ef19" />
> 
> <img width="1705" height="819" alt="PC3" src="https://github.com/user-attachments/assets/b8abb8ef-5883-4b22-b251-ebe034cb6565" />
> 
> 
> 
> Smartphone :
> 
> <img width="184" height="400" alt="Mobile1" src="https://github.com/user-attachments/assets/29193c8c-dd4b-4951-b915-4e2f7dc3503c" /> <img width="184" height="400" alt="Mobile2" src="https://github.com/user-attachments/assets/5005c2bd-373e-4b4c-9b4e-2e4df1c473ac" /> <img width="184" height="400" alt="Mobile3" src="https://github.com/user-attachments/assets/275ec48d-b25f-45f7-9efc-6d6fde6f22be" />
> 
> 
> 
> That’s actually a great idea : moving the selector into the Chat Form feels way more natural now that I’ve tested it 😄
> 
> The layout is cleaner, lighter, and it fits perfectly in the conversation flow. Definitely the right spot!
> 
> 
> 
> https://github.com/ggml-org/llama.cpp/pull/16562/commits/1324a4df92b0342a91fdc96660d1ae1704aa40ca

I will post a PR to improve the UI of this selector as now it's just taking too much space and looks a bit off 😜

ServeurpersoCom commented:
rebase master

ServeurpersoCom committed:
webui: make model selector more subtle and integrated into ChatForm

ServeurpersoCom commented:
<img width="873" height="783" alt="Sans titre" src="https://github.com/user-attachments/assets/f182065b-6fb4-4f56-a3aa-88b69f0d1324" />

We need a "developer" option with the model selector hidden by default,
but also a setting in lib/constants/settings-config.ts to keep it permanently enabled.
This is important so users can offer decentralized AI services to each other,
(like I'm doing right now with my friends)

ServeurpersoCom committed:
webui: replaced the Flowbite selector with a native Svelte dropdown

ServeurpersoCom committed:
webui: add developer setting to toggle the chat model selector

ServeurpersoCom commented:
PC :
https://github.com/user-attachments/assets/c7b55f8e-503d-4445-ab17-59ab4c75d35e

Mobile :
https://github.com/user-attachments/assets/f73c6e31-3246-4db0-b66a-75dba4d23b67

ServeurpersoCom committed:
webui: address review feedback from allozaur

Normalized streamed model names during chat updates
by trimming input and removing directory components before saving
or persisting them, so the conversation UI shows only the filename

Forced model names within the chat form selector dropdown to render as
a single-line, truncated entry with a tooltip revealing the full name

allozaur commented:
@ngxson can you please take a look at this PR and share your thoughts?

ServeurpersoCom committed:
webui: toggle displayed model source for legacy vs OpenAI-Compat modes

When the selector is disabled, it falls back to the active server model name from /props

When the model selector is enabled, the displayed model comes from the message metadata
(the one explicitly selected and sent in the request)

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatForm/ChatFormActions.svelte:
```diff
@@ -26,32 +28,35 @@
 		onMicClick,
 		onStop
 	}: Props = $props();
+	let currentConfig = $derived(config());
```
```suggestion
	}: Props = $props();

	let currentConfig = $derived(config());
```

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatForm/ChatFormModelSelector.svelte:
```diff
@@ -0,0 +1,399 @@
+			document.removeEventListener('pointerdown', handlePointerDown);
+			document.removeEventListener('keydown', handleKeydown);
+		};
+	});
```
This should be handled a bit differently:

```svelte
<script lang="ts">    
    onMount(async () => {
        try {
            await fetchModels();
        } catch (error) {
            console.error('Unable to load models:', error);
        } finally {
            isMounted = true;
        }
    });

    function handlePointerDown(event: PointerEvent) {
        if (!container) return;

        const target = event.target as Node | null;

        if (target && !container.contains(target) && !(menuRef && menuRef.contains(target))) {
            closeMenu();
        }
    }

    function handleKeydown(event: KeyboardEvent) {
        if (event.key === 'Escape') {
            closeMenu();
        }
    }
</script>

<!-- Svelte handles cleanup automatically -->
<svelte:document 
    onpointerdown={handlePointerDown}
    onkeydown={handleKeydown}
/>
```

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatForm/ChatFormModelSelector.svelte:
```diff
@@ -0,0 +1,399 @@
+				}
+			}
+		};
+	}
```
This could be externalized as it's own file in `$lib/utils`

allozaur commented on tools/server/webui/src/lib/stores/chat.svelte.ts:
```diff
@@ -359,28 +358,45 @@ class ChatStore {
+						console.error('Failed to persist model name:', error);
+						modelPersisted = false;
+					}
+				);
```
Let's externalize this as `$lib/utils/model-names.ts` where we'd have this utility function:

```ts
// lib/utils/model-names.ts

/**
 * Normalizes a model name by extracting the filename from a path.
 * 
 * Handles both forward slashes (/) and backslashes (\) as path separators.
 * If the model name is just a filename (no path), returns it as-is.
 * 
 * @param modelName - The model name or path to normalize
 * @returns The normalized model name (filename only)
 * 
 * @example
 * normalizeModelName('models/llama-3.1-8b') // Returns: 'llama-3.1-8b'
 * normalizeModelName('C:\\Models\\gpt-4') // Returns: 'gpt-4'
 * normalizeModelName('simple-model') // Returns: 'simple-model'
 * normalizeModelName('  spaced  ') // Returns: 'spaced'
 * normalizeModelName('') // Returns: ''
 */
export function normalizeModelName(modelName: string): string {
    const trimmed = modelName.trim();
    
    if (!trimmed) {
        return '';
    }
    
    const segments = trimmed.split(/[\\/]/);
    const candidate = segments.pop();
    const normalized = candidate?.trim();

    return normalized && normalized.length > 0 ? normalized : trimmed;
}

/**
 * Validates if a model name is valid (non-empty after normalization).
 * 
 * @param modelName - The model name to validate
 * @returns true if valid, false otherwise
 */
export function isValidModelName(modelName: string): boolean {
    return normalizeModelName(modelName).length > 0;
}
```

> [!IMPORTANT]
> Let's add a Vitest unit test for this.
>
> e. g.
>
> ```ts
> // lib/utils/model-names.test.ts
> import { describe, it, expect } from 'vitest';
> import { normalizeModelName, isValidModelName } from './model-names';
> 
> describe('normalizeModelName', () => {
>     it('extracts filename from forward slash path', () => {
>         expect(normalizeModelName('models/model-name-1')).toBe('model-name-1');
>         expect(normalizeModelName('path/to/model/model-name-2')).toBe('model-name-2');
>     });
>
>     it('extracts filename from backslash path', () => {
>         expect(normalizeModelName('C:\\Models\\model-name-1')).toBe('model-name-1');
>         expect(normalizeModelName('path\\to\\model\\model-name-2')).toBe('model-name-2');
>     });
>
>     it('handles mixed path separators', () => {
>         expect(normalizeModelName('path/to\\model/model-name-2')).toBe('model-name-2');
>     });
>
>     it('returns simple names as-is', () => {
>         expect(normalizeModelName('simple-model')).toBe('simple-model');
>         expect(normalizeModelName('model-name-2')).toBe('model-name-2');
>     });
>
>     it('trims whitespace', () => {
>         expect(normalizeModelName('  model-name
> ```

and then we can

```suggestion
		let resolvedModel: string | null = null;
		let modelPersisted = false;

		const recordModel = (modelName: string, persistImmediately = true): void => {
			const normalizedModel = normalizeModelName(modelName);

            if (!normalizedModel || normalizedModel === resolvedModel) {
                return;
            }

            resolvedModel = normalizedModel;

            const messageIndex = this.findMessageIndex(assistantMessage.id);

            this.updateMessageAtIndex(messageIndex, { model: normalizedModel });

            if (persistImmediately && !modelPersisted) {
                modelPersisted = true;
                DatabaseStore.updateMessage(assistantMessage.id, { model: normalizedModel }).catch(
                    (error) => {
                        console.error('Failed to persist model name:', error);
                        modelPersisted = false;
                        resolvedModel = null; // Reset to allow retry
                    }
                );
```

ServeurpersoCom replied:
```
 *
 * @example
 * normalizeModelName('models/llama-3.1-8b') // Returns: 'llama-3.1-8b'
 * normalizeModelName('C:\\Models\\gpt-4') // Returns: 'gpt-4'
 * normalizeModelName('simple-model') // Returns: 'simple-model'
 * normalizeModelName('  spaced  ') // Returns: 'spaced'
 * normalizeModelName('') // Returns: ''
```

allozaur reviewed (changes requested):
I think we can improve the architecture of this and it's better do have this implemented properly from the start rahter than need to clean it up in the future.

ServeurpersoCom committed:
Update tools/server/webui/src/lib/components/app/chat/ChatForm/ChatFormActions.svelte

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/constants/localstorage-keys.ts

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/components/app/chat/ChatForm/ChatFormModelSelector.svelte

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/components/app/chat/ChatMessages/ChatMessageAssistant.svelte

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/services/chat.ts

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/services/chat.ts

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
webui: refactor model selector and persistence helpers

- Replace inline portal and event listeners with proper Svelte bindings
- Introduce 'persisted' store helper for localStorage sync without runes
- Extract 'normalizeModelName' utils + Vitest coverage
- Simplify ChatFormModelSelector structure and cleanup logic

Replaced the persisted store helper's use of '$state/$effect' runes with
a plain TS implementation to prevent orphaned effect runtime errors
outside component context

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
webui: document normalizeModelName usage with inline examples

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatForm/ChatFormModelSelector.svelte:
```diff
@@ -0,0 +1,357 @@
+
+	$effect(() => {
+		const optionCount = options.length;
+		if (!isOpen || optionCount < 0) return;
```
```suggestion
		const optionCount = options.length;

		if (!isOpen || optionCount < 0) return;
```

allozaur reviewed (changes requested):
@ServeurpersoCom few cosmetic changes left to address

ServeurpersoCom committed:
Update tools/server/webui/src/lib/components/app/chat/ChatForm/ChatFormModelSelector.svelte

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/stores/models.svelte.ts

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/stores/models.svelte.ts

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
webui: extract ModelOption type into dedicated models.d.ts

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
webui: refine ChatMessageAssistant displayedModel source logic

ServeurpersoCom committed:
webui: stabilize dropdown, simplify model extraction, and init assistant model field

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatMessages/ChatMessageAssistant.svelte:
```diff
@@ -70,6 +71,22 @@
+
+		return serverModel;
+	});
+	const handleCopyModel = () => {
```
```suggestion
	});

	function handleCopyModel() {
```

ServeurpersoCom committed:
chore: update webui static build

allozaur reviewed (changes requested):
Also please make this small change ;)

ServeurpersoCom committed:
Update tools/server/webui/src/lib/components/app/chat/ChatMessages/ChatMessageAssistant.svelte

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
chore: npm format, update webui static build

ServeurpersoCom committed:
webui: align sidebar trigger position, remove z-index glitch

ServeurpersoCom committed:
chore: update webui build output

allozaur reviewed (approved):
Okay, i think that we are good to go :)
